src/kqueue.rs

```diff
@@ -0,0 +1,380 @@
+                                could only emulate this inotify behavior if we keep track of
+                                all files and directories internally and then perform a
+                                lookup.
+                                */
+                                kqueue::Vnode::Delete => {
+                                    remove_watches.push(path.clone());
+                                    Event::new(EventKind::Remove(RemoveKind::Any))
+                                }
+
+                                //data was write to this file
+                                kqueue::Vnode::Write => Event::new(EventKind::Access(
+                                    AccessKind::Close(AccessMode::Write),
+                                )),
+
+                                /*
+                                Extend and Truncate are just different names for the same
+                                operation, extend is only used on FreeBSD, truncate everwhere
+                                else
+                                */
+                                kqueue::Vnode::Extend => Event::new(EventKind::Modify(
+                                    ModifyKind::Data(DataChange::Size),
+                                )),
+                                kqueue::Vnode::Truncate => Event::new(EventKind::Modify(
+                                    ModifyKind::Data(DataChange::Size),
+                                )),
+
+                                /*
+                                this kevent has the same problem as the delete kevent. The
+                                only way i can think of providing "better" event with more
+                                information is to do the diff our self, while this maybe do
+                                able of delete. In this case it would somewhat expensive to
+                                keep track and compare ever peace of metadata for every file
+                                */
+                                kqueue::Vnode::Attrib => Event::new(EventKind::Modify(
+                                    ModifyKind::Metadata(MetadataKind::Any),
+                                )),
+
+                                /*
+                                The link count on a file changed => subdirectory created or
+                                delete. Currently now idea how to track this.
+                                */
+                                kqueue::Vnode::Link => {
+                                    // readd this folder, this is currently the easiest way to
+                                    // track new and removed subfolders
+
+                                    // TODO: This is really expensive, as we recursively walk through
+                                    // all subdirectories.
+                                    remove_watches.push(path.clone());
+                                    add_watches.push(path.clone());
```

Comment:
@worr do you know of a better way than to re-index the whole tree ?

Reply:
We could store a HashMap of all know paths and then compare with paths are missing or where added to the directory.
I'll try to implement this.

Reply:
After thinking about this a bit, i cam to the conclusion that a simple `HashSet<PathBuf>` won't cut it. As i don't think there is any way to get from the directory we get from `NOTE_LINK` to the files we expect to be in that directory.

I think we need a `HashMap<PathBuf, Vec<PathBuf>>` where the key is the directory and the value is a list of all files in the directory at time of adding. This feels like a race condition to me, as files could be created of deleted will we are setting up this directory watch. Not quite sure if this is important in this case.

Reply:
@0xpr03 We've been chatting about this out-of-band a bit. `kqueue(2)` doesn't expose this in a meaningful way, and this is especially expensive since we need to open file handles to each file on index (this is a requirement of `kqueue(2)`). The method @xanderio brought up is one of the ones we've been talking about.

Reply:
Hm, thanks for investigating. I'm fine with the current solution, it's not the first time where we have to re-index due to race conditions. Could you change the todo so it documents why we have to use this approach ?

Reply:
I'll can do that, all though I personally would prefer if  we left the TODO in places where there are thinks that we see that can possibly be improve in the future, but we just don't know the solution to at the moment. But that's just my 2ct

Reply:
on a second thought yeah, you can leave it like that